### PR TITLE
KEYCLOAK-10373 Migrate Keycloak for SSO

### PR DESCRIPTION
**Pull request by andymunro**
@stianst Can you check this works?

### CODE REVIEW

File: upgrading/topics/rhsso/intro.adoc [hunk 1]
```diff
@@ -14,7 +14,7 @@ RH-SSO instances from a single control point. The upgrade process differs depend
 implemented. Specific instructions for each mode are provided where applicable.
 
 The purpose of this guide is to document the steps that are required to successfully upgrade from
-Red Hat Single Sign-On 7.2 to Red Hat Single Sign-On 7.3.
+Red Hat Single Sign-On 7.2 or Keycloak 4.8.x to Red Hat Single Sign-On 7.3.
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: I would remove this. It's sufficient with the separate section on migrating from Keycloak.

File: upgrading/topics/rhsso/intro.adoc [hunk 2]
```diff
@@ -31,6 +31,8 @@ fixes, and new features. If you plan to upgrade from one Red Hat Single Sign-On
 from Red Hat Single Sign-On 7.2 to Red Hat Single Sign-On 7.3, code changes should not be required for applications or
 custom server extensions as long as no private, unsupported, or tech preview APIs are used.
 
+The steps for minor updates also apply for migrating from Keycloak to Red Hat Single Sign-On.
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: Remove this.

File: upgrading/topics/rhsso/intro.adoc [hunk 3]
```diff
@@ -41,3 +43,20 @@ installations can also rollback the patch and server configuration.
 A micro release only contains the artifacts that have changed. For example if Red Hat Single Sign-On 7.2.1 contains changes to
 the server and the JavaScript adapter, but not the EAP adapter, only the server and JavaScript adapter are released and require
 updating.
+
+=== Migration from Keycloak
+
+You can migrate to Red Hat Single Sign-On, the supported Red Hat product, from Keycloak, the community project. Some differences exist in the installed software, such as the following:
+
+* Red Hat Single Sign-On uses JBoss EAP in place of WildFly for the built-in application server
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: Either remove this, or it needs more explanation. Strictly speaking it's an implementation detail.

File: upgrading/topics/rhsso/intro.adoc [hunk 4]
```diff
@@ -41,3 +43,20 @@ installations can also rollback the patch and server configuration.
...
+* Some Keycloak features are not included in Red Hat Single Sign-On, such as some of the Service Provider Interfaces (SPIs).
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: Remove this. This is not correct. SPIs that are supported in community are also supported in product. SPIs that are not supported are available in both community and product.

File: upgrading/topics/rhsso/intro.adoc [hunk 5]
```diff
@@ -41,3 +43,20 @@ installations can also rollback the patch and server configuration.
...
+
+.Prerequisites
+
+* To learn about new features before the upgrade, review the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.3/html-single/upgrading_guide/index#changes[changes].
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: Don't hard-code links

File: topics/templates/document-attributes-product.adoc [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 :project_name_full: Red Hat Single Sign-On
 :project_version: 7.4.0.GA
+:keycloak_upgrade_version: 9.0.0
```
Comment (stianst) [inline topics/templates/document-attributes-product.adoc]: 9.0.x

File: upgrading/topics/rhsso/intro.adoc [hunk 7]
```diff
@@ -41,3 +43,20 @@ installations can also rollback the patch and server configuration.
...
+* Verify that you have installed the correct version of Keycloak as a starting point. To migrate to Red Hat Single Sign-On {project_versionDoc}, you need Keycloak {keycloak_upgrade_version}.
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc:52]: you need Keycloak {keycloak_upgrade_version} or older

File: upgrading/topics/rhsso/intro.adoc [hunk 8]
```diff
@@ -41,3 +43,20 @@ installations can also rollback the patch and server configuration.
...
+
+.Procedure
+
+. Perform the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.3/html-single/upgrading_guide/index#upgrading_minor[Minor Upgrade procedure]. Although this procedure is labelled *Minor Upgrade*, the same steps apply for this migration.
```
Comment (stianst) [inline upgrading/topics/rhsso/intro.adoc]: Remove hard-coded links

**Review by stianst — CHANGES_REQUESTED** [review]

File: upgrading/topics/rhsso/intro.adoc (hunk 7 above)
Comment (andymunro) [reply upgrading/topics/rhsso/intro.adoc:52]: @stianst  I understood from Vlasta that users would need a 9.0.x version.  See:  https://issues.redhat.com/browse/KEYCLOAK-10373
Are we relaxing this prerequisite to allow earlier versions?

File: upgrading/topics/rhsso/intro.adoc (hunk 8 above)
Comment (andymunro) [reply upgrading/topics/rhsso/intro.adoc]: Should 7.3 become {project_versionDoc}?

File: upgrading/topics/rhsso/intro.adoc (hunk 4 above)
Comment (andymunro) [reply upgrading/topics/rhsso/intro.adoc]: Okay, I have removed it.  However, I know we only document the User Storage SPI in SSO.  Is the difference that do not document all SPIs in SSO?

**Comment by andymunro** [comment]
@stianst Do you have any more changes to request?

**Review by hmlnarik — APPROVED** [review]
IMO, the remaining comment can be addressed post GA if needed.

**Review by vramik — APPROVED** [review]

**Pull request merged**